src/common.ts

```diff
@@ -1,84 +1,74 @@
-import type { Client, OperationResult } from '@urql/core'
-import { atom } from 'jotai/vanilla'
+import type { Client, OperationContext, OperationResult } from '@urql/core'
+import { AnyVariables, Operation } from '@urql/core'
+import { atomWithObservable } from 'jotai/utils'
 import type { Getter } from 'jotai/vanilla'
-import { atomWithObservable } from 'jotai/vanilla/utils'
-import { filter, pipe, toObservable } from 'wonka'
+import { atom } from 'jotai/vanilla'
 import type { Source } from 'wonka'
-
-export const createAtoms = <
-  Args,
-  Result extends OperationResult,
-  Action,
-  ActionResult
->(
+import { pipe, toObservable } from 'wonka'
+import { suspenseAtom } from './suspenseAtom'
+
+export type InitialOperationResult<Data, Variables extends AnyVariables> = Omit<
+  OperationResult<Data, Variables>,
+  'operation'
+> & {
+  operation: Operation<Data, Variables> | undefined
+}
+// This is the same (aside from missing fetching and having hasNext) object shape as urql-react has by default while operation is yet to be triggered/yet to be fetched
+export const urqlReactCompatibleInitialState = {
+  stale: false,
+  // Casting is needed to make typescript chill here as it tries here to be too smart
+  error: undefined as any,
+  data: undefined as any,
+  extensions: undefined as any,
+  hasNext: false,
+  operation: undefined,
+} as InitialOperationResult<any, any>
+
+export const createAtoms = <Args, Result extends OperationResult, ActionResult>(
   getArgs: (get: Getter) => Args,
   getClient: (get: Getter) => Client,
   execute: (client: Client, args: Args) => Source<Result>,
-  handleAction: (
-    action: Action,
-    client: Client,
-    refresh: () => void
-  ) => ActionResult
+  reexecute: (context: Partial<OperationContext>, get: Getter) => ActionResult,
+  getPause: (get: Getter) => boolean
 ) => {
-  const refreshAtom = atom(0)
-
-  if (process.env.NODE_ENV !== 'production') {
-    refreshAtom.debugPrivate = true
-  }
-
   const sourceAtom = atom((get) => {
-    get(refreshAtom)
     const args = getArgs(get)
     const client = getClient(get)
-    const source = execute(client, args)
+    const source = getPause(get) ? null : execute(client, args)
     return source
   })
 
   if (process.env.NODE_ENV !== 'production') {
     sourceAtom.debugPrivate = true
   }
 
-  const baseStatusAtom = atom((get) => {
-    const source = get(sourceAtom) as Source<Result | undefined>
-    const observable = pipe(source, toObservable)
-    const resultAtom = atomWithObservable(() => observable, {
-      initialValue: undefined,
-    })
-
-    if (process.env.NODE_ENV !== 'production') {
-      resultAtom.debugPrivate = true
-    }
+  const initialLoadAtom = atom<Result>(
+    urqlReactCompatibleInitialState as Result
+  )
 
-    return resultAtom
+  const cacheResultAtom = atom<{ cache: Map<number, Result> }>({
+    cache: new Map(),
```

Comment:
Again, this is very tricky, but `new Map()` is called during the definition, so it's shared across providers.

Reply:
Will add to onMount. Doesn't really matter. But to clarify why this is needed.

This is hack to preserve a cache of all state changes (operation results that are getting piped/streamed from source) of the query, and then show the last one, when we pause (getPause returns true) and our source becomes null (so we unsubscribe from changes).

And so instead of undefined or initial state placeholder, we show last state before the pause. In case of unpause we will subscribe to the source again, and show actual data from the urql client cache that will be piped from it through source into our observable.

Reply:
Changed that to a simple object